presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OriginalFilesUtils.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.orc;
+
+import io.prestosql.orc.OrcDataSource;
+import io.prestosql.orc.OrcDataSourceId;
+import io.prestosql.orc.OrcReader;
+import io.prestosql.orc.OrcReaderOptions;
+import io.prestosql.plugin.hive.FileFormatDataSourceStats;
+import io.prestosql.plugin.hive.HdfsEnvironment;
+import io.prestosql.spi.PrestoException;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.util.Collection;
+
+import static io.prestosql.plugin.hive.AcidInfo.OriginalFileInfo;
+import static io.prestosql.plugin.hive.HiveErrorCode.HIVE_CANNOT_OPEN_SPLIT;
+
+public final class OriginalFilesUtils
+{
+    private OriginalFilesUtils() {}
+
+    /**
+     * Returns total number of rows present before the given original file in the same bucket.
+     * example: if bucket-1 has original files
+     * 000000_0 -> X0 rows
+     * 000000_0_copy1 -> X1 rows
+     * 000000_0_copy2 -> X2 rows
+     * <p>
+     * for 000000_0_copy2, it returns (X0+X1)
+     */
+    public static long getPrecedingRowCount(
+            Collection<OriginalFileInfo> originalFileInfos,
+            Path splitPath,
+            HdfsEnvironment hdfsEnvironment,
+            String sessionUser,
+            OrcReaderOptions options,
+            Configuration configuration,
+            FileFormatDataSourceStats stats)
+    {
+        long rowCount = 0;
+        for (OriginalFileInfo originalFileInfo : originalFileInfos) {
```

Comment:
It feels like we are doing `O(N^2)` FS operations. Where `N` is number of original files? Is that a problem? Should we cash already computed results of `getRowsInFile`? Or compute `PreceedingRowCounts` for all original files in one go and then just use the result?

Reply:
If we precompute the rows in master node, then:
1. scheduling time will increase as master has to locate, read and calculate the row id.
2. we are doing distributing such work to workers.
3. we need to add mechanism to read original files such as creating orc reader.

If we keep a map, then:
1. map has to be evicting cache to remove the stale entry and we don't know the eviction time.
2. we have to store the complete path of an original file as a key, which may put pressure on memory.
3. in case original file split gets scheduled to a different worker, we need to add calculation again.

So, considering all such cases, we thought of going ahead with simple computation. We can further discuss this to come up with an optimal solution.

Reply:
Yeah I think this can be a followup. I just feel not comfortable to leave it as is. I would not like to start bombarding name-node with queries because of some pathological ACID table with 1000 original files. 1000 is not that many.
I do not expect memory pressure to be something to worry about if we setup max cache size to something reasonable.

Reply:
If we can come up with a cache size and eviction time, then we can do this too.

Reply:
I think that instead of a cache, we should do this linearly as part of the split generation process.
The downside is that this moves the footer access to the coordinator. However, there are some 
other advantages to be derived from footer access on the coordinator as well.

For now, I am adding a TODO comment in the code.

Reply:
Moving original file footer read part to coordinate may have a severe impact on schedule time. In a production environment, tables generally have a large number of original files. Say, we have 20K original files, we will be reading footer and calculating start row ID for each of the original files. Plus, we need to move ORC read stuff too.

Another solution we can do it to maintain a query/task level cache of the row ID computed from footer and clear the cache after the query/task finishes. This way we will be delegating work in a distributed manner and speed up. It may be added in the PR. thoughts?

Reply:
> Say, we have 20K original files, we will be reading footer and calculating start row ID for each of the original files. Plus, we need to move ORC read stuff too

in such a case current approach results in 200M ORC footer accesses to calculate start row IDs, where as 20K could be enough.
Perhaps, I am missing something obvious.

Reply:
We are optimising for worst case of thousands of file under same bucket if we are considering complexity of O(n^2). In average case there would be much better distribution and not all nodes will have to read all files as only the list of files  in the same bucket, lexicographically coming before the file for this split, are passed down here.
A cache which provides singleton load of key, worst case would be O(n) and given different splits will load different files it should be better than loading all files in master first.
It should be simple to try to different logics and pick the best one, separately from this PR.